### PR TITLE
Use generated functions instead of @ngenerate macro

### LINKED ISSUE #269: Get rid of @ngenerate
This is a relic from Julia versions past, available only through Compat. We can ditch the Compat dependency altogether if we can remove the use of `@ngenerate`, which is currently only used for `mapreducedim`.

### PR DESCRIPTION
**Pull request by andreasnoack**
Fixes #269

### CODE REVIEW

File: src/reducedim.jl [hunk 1]
```diff
@@ -5,23 +5,25 @@ using Base.check_reducedims
...
-    @nloops(N, i, A,
-        d->(state_{d-1} = state_d),
-        d->(skip_d || (state_d = state_0)),
-        begin
-            @inbounds R[state_0] &= f(Base.unsafe_bitgetindex(Achunks, k))
-            state_0 += 1
-            k += 1
-        end)
-    R
+@generated function Base._mapreducedim!{N}(f, op::typeof(&), R::Array{Bool}, A::BitArray{N})
```
Comment (ararslan) [inline src/reducedim.jl]: Can the `{N}` be removed in favor of a trailing `where N`?

**Comment by ararslan** [comment]
Well that's a _much_ simpler change than I was imagining. Thanks so much!

**Comment by ararslan** [comment]
Want to drop the Compat dependency while you're in here?

File: src/reducedim.jl [hunk 2]
```diff
@@ -5,23 +5,25 @@ using Base.check_reducedims
...
-    @nloops(N, i, A,
-        d->(state_{d-1} = state_d),
-        d->(skip_d || (state_d = state_0)),
-        begin
-            @inbounds R[state_0] &= f(Base.unsafe_bitgetindex(Achunks, k))
-            state_0 += 1
-            k += 1
-        end)
-    R
+@generated function Base._mapreducedim!(f, op::typeof(&), R::Array{Bool}, A::BitArray{N}) where {N}
```
Comment (tkelman) [inline src/reducedim.jl]: why is this here and not base? this is type piracy

Comment (andreasnoack) [inline src/reducedim.jl]: I guess this definition is older than the concept of type piracy. @simonster why didn't you submit this one to base?

Comment (tkelman) [inline src/reducedim.jl]: ping, is it still necessary to have this here?

Comment (ararslan) [inline src/reducedim.jl]: No, it should just be able to be removed here, though it would be good to know whether moving the code to Base would provide a performance improvement.

Comment (tkelman) [inline src/reducedim.jl]: then let's delete this
